### PR TITLE
[GEOT-5182] ShadedRelief implementation

### PR DESCRIPTION
**Pull request by etj**
Work in progress
- [x] tests are still missing
- [x] jai-ext implementation has still to be set as a dependency i.e.
  - [x] jt-shadedrelief PR needs to be merged
  - [x] jai-ext needs a new release - Done: `1.0.23`
  - [x] jai-ext version has to be change in GT

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/styling/SLDParser.java [hunk 1]
```diff
@@ -1554,11 +1554,11 @@ protected ShadedRelief parseShadedRelief(Node root) {
                 childName = child.getNodeName();
             }
             if ("BrightnessOnly".equalsIgnoreCase(childName)) {
-                symbol.setBrightnessOnly(Boolean.getBoolean(getFirstChildValue(child)));
+                symbol.setBrightnessOnly(Boolean.parseBoolean(getFirstChildValue(child)));
```
Comment (bradh) [inline modules/library/main/src/main/java/org/geotools/styling/SLDParser.java]: This part of the PR (on SLD parsing) looks unrelated. Is it meant to be included?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ShadedReliefNode.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        registry.registerDescriptor(op);
+        String descName = op.getName();
+
+        // registering the RIF
+        RenderedImageFactory rif = new ShadedReliefRIF();
+        registry.registerFactory(
+                RenderedRegistryMode.MODE_NAME, descName, "org.geotools.gce.processing", rif);
+    }
+
+    private boolean brigthnessOnly = false;
```
Comment (bradh) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ShadedReliefNode.java]: Probably should be `brightnessOnly`

File: modules/library/main/src/main/java/org/geotools/styling/SLDParser.java (hunk 1 above)
Comment (etj) [reply modules/library/main/src/main/java/org/geotools/styling/SLDParser.java]: `BrightnessOnly` is part of the shadedRelief specs.
The `getBoolean()` call is clearly a typo, since it retrieves the value of a system property, while the `parseBoolean()` will consider the text itself as a string to be parsed to a boolean value.

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ShadedReliefNode.java (hunk 2 above)
Comment (etj) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ShadedReliefNode.java]: Yep, a typo, anyway it's private and won't be leaked outside. Thanks for spotting it!

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/RasterSymbolizerHelper.java [hunk 3]
```diff
@@ -225,47 +226,92 @@ public void visit(RasterSymbolizer rs) {
...
+
+        final ChannelSelectionNode csNode = new ChannelSelectionNode();
+        currNode = csNode;
+
+        currNode.addSource(prevNode);
+        prevNode = currNode;
+
         final ChannelSelection cs = rs.getChannelSelection();
-        csNode.addSource(this.getSource(0));
-        csNode.addSink(cmNode);
```
[the remark this answers is not in the document]
Comment (etj) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/RasterSymbolizerHelper.java]: Every node block has this assignments:
```
        currNode.addSource(prevNode);
        prevNode.addSink(currNode);
```

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/RasterSymbolizerHelper.java [hunk 4]
```diff
@@ -225,47 +226,92 @@ public void visit(RasterSymbolizer rs) {
...
+        boolean applyShadedRelief = !Double.isNaN(srNode.getReliefFactor());
+        boolean applyContrastEnhancement = ceNode.getType() != null;
+        if (applyShadedRelief && applyContrastEnhancement) {
+            throw new IllegalArgumentException(
+                    "ContrastEnhancement and ShadedRelief can't be applied at the same time. ");
+        }
+
+        if (applyShadedRelief) {
+
+            currNode.addSource(prevNode);
```
[the remark this answers is not in the document]
Comment (etj) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/RasterSymbolizerHelper.java]: Ok, previous code seemed quite confused about this chain. Will fix.

**Comment by dromagnoli** [comment]
Note that jai-ext has been moved to 1.1.1 these days. You may want to update the pom

**Comment by etj** [comment]
Merged with master, so the changes in the pom file are automatically in.
